src/fluree/db/ledger/transact/tempid.clj

```diff
@@ -0,0 +1,136 @@
+                      {:status 400
+                       :error  :db/invalid-tx})))))
+
+
+(defn set
+  "Sets a tempid value into the cache. If the tempid was already set by another :upsert
+  predicate that matched a different subject, throws an error. Returns set subject-id on success."
+  [tempid subject-id {:keys [tempids] :as tx-state}]
+  (swap! tempids update tempid
+         (fn [existing-sid]
+           (cond
+             (nil? existing-sid) subject-id                 ;; hasn't been set yet, success.
+             (= existing-sid subject-id) subject-id         ;; resolved, but to the same id - ok
+             :else (throw (ex-info (str "Temp-id " (:user-string tempid)
+                                        " matches two (or more) subjects based on predicate upserts: "
+                                        existing-sid " and " subject-id ".")
+                                   {:status 400
+                                    :error  :db/invalid-tx})))))
+  subject-id)
+
+
+(defn result-map
+  "Creates a map of original user tempid strings to the resolved value."
+  [{:keys [tempids] :as tx-state}]
+  (reduce-kv (fn [acc ^TempId tempid subject-id]
+               (if (:unique? tempid)
+                 (assoc acc (:user-string tempid) subject-id)
+                 (update acc (:user-string tempid) (fn [[min-sid max-sid]]
+                                                     [(if min-sid (min min-sid subject-id) subject-id)
+                                                      (if max-sid (max max-sid subject-id) subject-id)]))))
+             {} @tempids))
+
+
+(defn flake
+  "Returns flake for tx-meta (transaction sid) that contains a json packaging of the tempids map."
+  [tempids-map t]
+  (flake/->Flake t const/$_tx:tempids (json/stringify tempids-map) t true nil))
+
+(defn assign-subject-ids
+  "Assigns any unresolved tempids with a permanent subject id."
+  [{:keys [tempids tempids-ordered upserts db-before t] :as tx-state} tx]
+  (try
+    (let [ecount      (assoc (:ecount db-before) -1 t)      ;; make sure to set current _tx ecount to 't' value, even if no tempids in transaction
+          tempids-map @tempids]
+      (loop [[tempid & r] @tempids-ordered
+             ;[[tempid resolved-id] & r] @tempids
+             tempids-map* tempids-map
+             upserts*     #{}
+             ecount*      ecount]
+        (if (nil? tempid)                                   ;; finished
+          (do (reset! tempids tempids-map*)
+              (when-not (empty? upserts*)
+                (reset! upserts upserts*))
+              ;; return tx-state, don't need to update ecount in db-after, as dbproto/-with will update it
+              tx-state)
+          (if (nil? (get tempids-map tempid))
+            (let [cid      (dbproto/-c-prop db-before :id (:collection tempid))
+                  next-id  (if (= -1 cid)
```

Comment:
do we have a constant or mapping somewhere that makes it explicit what `-1` signifies as a cid? if so, i think we should use that here instead of the raw number.

Reply:
Good idea, adding it to fluree/db constants